flask_monitoringdashboard/controllers/requests.py

```diff
@@ -30,6 +35,74 @@ def get_num_requests_data(db_session, start_date, end_date):
     }
 
 
+def get_all_request_status_code_counts(db_session, endpoint_id):
+    """
+    Gets all the request status code counts.
+
+    :param db_session: session for the database
+    :param endpoint_id: id for the endpoint
+    :return: A list of tuples in the form of `(status_code, count)`
+    """
+    return db_session.query(Request.status_code, func.count(Request.status_code)) \
+        .filter(and_(Request.endpoint_id == endpoint_id, Request.status_code.isnot(None))) \
+        .group_by(Request.status_code).all()
+
+
+def get_status_code_distribution(db_session, endpoint_id):
+    """
+    Gets the distribution of status codes returned by the given endpoint.
+
+    :param db_session: session for the database
+    :param endpoint_id: id for the endpoint
+    :return: A dict where the key is the status code and the value is the fraction of requests that returned the status
+    code. Example: a return value of `{ 200: 0.92, 404: 0.08 }` means that status code 200 was returned on 92% of the
+    requests. 8% of the requests returned a 404 status code.
+    """
+    status_code_counts = get_all_request_status_code_counts(db_session, endpoint_id)
+
+    total_count = sum(frequency for (_, frequency) in status_code_counts)
+
+    return {status_code: frequency / total_count for (status_code, frequency) in status_code_counts}
+
+
+def get_status_code_frequencies(db_session, endpoint_id):
+    """
+    Gets the frequencies of each status code.
+
+    :param db_session: session for the database
+    :param endpoint_id: id for the endpoint
+    :return: A dict where the key is the status code and the value is the fraction of requests that returned the status
+    code. Example: a return value of `{ 200: 105, 404: 3 }` means that status code 200 was returned 105 times and
+    404 was returned 3 times.
+    """
+    status_code_counts = db_session.query(Request.status_code, func.count(Request.status_code)) \
+        .filter(and_(Request.endpoint_id == endpoint_id, Request.status_code.isnot(None))) \
+        .group_by(Request.status_code).all()
+
+    return dict(status_code_counts)
+
+
+def get_error_requests(db_session, endpoint_id, *criterion):
+    """
+    Gets all requests that did not return a 200 status code.
+
+    :param db_session:
+    :param endpoint_id:
+    :param criterion:
+    :return:
+    """
+    return db_session \
```

Comment:
@thejager it's very likely you knew, but you can split something over multiple lines if you put the expression between parentheses: that way you don't have to terminate lines with \

Reply:
I didn't know actually, will change this soon 😎